utils/ipfs.py

```diff
@@ -0,0 +1,241 @@
+#  IPFS has two CID formats v0 and v1, v1 supports different encodings, defaults are:
+#  CIDv0:
+#    base58btc              Qm ✓ QmRKs2ZfuwvmZA3QAWmCqrGUjV9pxtBUDP3wuc6iVGnjA2
+#  CIDv1:
+#    base16                  f ✓ f017012202c5f688262e0ece8569aa6f94d60aad55ca8d9d83734e4a7430d0cff6588ec2b
+#    base16upper             F ✓ F017012202C5F688262E0ECE8569AA6F94D60AAD55CA8D9D83734E4A7430D0CFF6588EC2B
+#    base32upper             B ✓ BAFYBEIBML5UIEYXA5TUFNGVG7FGWBKWVLSUNTWBXGTSKOQYNBT7WLCHMFM
+#    base32                  b ✓ bafybeibml5uieyxa5tufngvg7fgwbkwvlsuntwbxgtskoqynbt7wlchmfm
+#    base58btc               z ✓ zdj7WYR7PzjmRQNRsMKuFipiE73MhMGgRbc5hTUaQVPJiMdKx
+#    base64                  m ✓ mAXASICxfaIJi4OzoVpqm+U1gqtVcqNnYNzTkp0MNDP9liOwr
+#    base64url               u ✓ uAXASICxfaIJi4OzoVpqm-U1gqtVcqNnYNzTkp0MNDP9liOwr
+#    base64urlpad            U ✓ UAXASICxfaIJi4OzoVpqm-U1gqtVcqNnYNzTkp0MNDP9liOwr
+#
+
+REG_CID_0_58_BTC = r"Qm[1-9A-HJ-NP-Za-km-z]{44,128}"
+REG_CID_1_16 = r"[fF][0-9a-zA-F]{50,128}"  # case-insensitive
+REG_CID_1_32 = r"[bB][a-zA-Z2-7]{58,128}"  # case-insensitive
+REG_CID_1_58_BTC = r"z[1-9A-HJ-NP-Za-km-z]{48,128}"
+REG_CID_1_64 = r"m[+A-Za-z0-9/]{40,128}"
+REG_CID_1_64_URL = r"u[-A-Za-z0-9_]{40,128}"
+REG_CID_1_64_URLPAD = r"U[-A-Za-z0-9_]{40,128}={0,3}"
+
+REG_CID_DEFAULT = rf"\b({REG_CID_0_58_BTC}|{REG_CID_1_16}|{REG_CID_1_32}|{REG_CID_1_58_BTC}|{REG_CID_1_64}|{REG_CID_1_64_URL}|{REG_CID_1_64_URLPAD})\b"
+REG_ETH_ADDRESS = r"\b(0x[a-fA-F0-9]{40})\b"
+
+REG_VOTE_CID = rf"\b({REG_CID_1_32})\b"
+VOTE_CID_PREFIX = "lidovoteipfs://"
+REG_VOTE_CID_WITH_PREFIX_LAST = rf"\b{VOTE_CID_PREFIX}{REG_VOTE_CID}\s*$"
+
+
+class IPFSUploadResult(TypedDict):
+    cid: str
+    text: str
+    messages: list[Tuple[str, str]]
+
+
+# alternative for upload_str_to_web3_storage
+def _upload_str_to_infura_io(text: str) -> str:
+    text_bytes = text.encode("utf-8")
+    text_file = io.BytesIO(text_bytes)
+    files = {"file": text_file}
+    (projectId, projectSecret) = get_infura_io_keys()
+
+    endpoint = "https://ipfs.infura.io:5001"
+
+    response = requests.post(endpoint + "/api/v0/add?cid-version=1", files=files, auth=(projectId, projectSecret))
+    response.raise_for_status()
+    response_json = response.json()
+
+    return response_json.get("Hash")
+
+
+# upload text to web3.storage ipfs
+def _upload_str_to_web3_storage(text: str) -> str:
+    text_bytes = text.encode("utf-8")
+    text_file = io.BytesIO(text_bytes)
+    web3_storage_token = get_web3_storage_token()
+
+    endpoint = "https://api.web3.storage/upload"
+    headers = {"Authorization": f"Bearer {web3_storage_token}", "Content-Type": "application/x-directory"}
+
+    response = requests.post(endpoint, headers=headers, data=text_file)
+    response.raise_for_status()
+    response_json = response.json()
+
+    return response_json.get("cid")
+
+
+# uploading text to ipfs
+def _upload_str_to_ipfs(text: str, service: str = "web3.storage") -> str:
```

Comment:
maybe we can detect the service based on env vars set

Reply:
alternative